include/mbedtls/ssl.h

```diff
@@ -1446,6 +1454,13 @@ struct mbedtls_ssl_config
 #if defined(MBEDTLS_DHM_C) && defined(MBEDTLS_SSL_CLI_C)
     unsigned int MBEDTLS_PRIVATE(dhm_min_bitlen);    /*!< min. bit length of the DHM prime   */
 #endif
+
+    /** User data pointer or handle.
+     *
+     * The library sets this to \p 0 when creating a context and does not
+     * access it afterwards.
+     */
+    uintptr_t MBEDTLS_PRIVATE(user_data);
```

Comment:
Care is taken in the interfaces to distinguish between an integer and a pointer thus why not rather using an union here and thus not having at all to think about the portability issues of integer vs pointer conversions in C ?

Reply:
Yes, makes sense.

I even wonder if the function interface should be a union. But it would make the functions harder to use: most users only ever care about storing a pointer, or only ever care about storing an integer.

Reply:
Actually, now that [I've coded it](https://github.com/gilles-peskine-arm/mbedtls/tree/ssl-get-version-3.1-union), I think that a union is more fragile, so I prefer to use a cast.

Of the assumptions we make about the C implementation, only two are relevant here:
* `uintptr_t` exists and has no padding bits.
* All-bits-zero as a pointer is a null pointer.

Consider a “weird”, but valid architecture with fat pointers that contain a lot of redundancy, and where `uintptr_t` is smaller than `void*`. Then initializing the union to 0 via its `uintptr_t` member does not fully initialize the part of the storage.
```
ssl.user_data.n = 0;
ssl.user.data.p; // undefined behavior: p is not initialized
```

Our code may still work because I think we only ever initialize the structures with `memset(p, 0, sizeof(*p))`, and _that_ allows reading the pointer as a null pointer. But it means `*p = 0` will not work, which is fragile.